### ISSUE TITLE
Release plan

### ISSUE THREAD

**Issue by Andersson007** [ansible-collections/ansible.mysql#31]
##### SUMMARY
_(partially copied from https://github.com/ansible-collections/community.crypto/issues/74 , thanks to @felixfontein)_

We should decide eventually on how to release this collection (w.r.t. versioning).
Small collections like this one don't need a complex plan like the one for community.general and community.network.
So how about the following?

1. Release minor and patch releases whenever we want (like after adding new features or fixing bugs). Since this collection is small, there's no need to fix things in advance. Just add features, and after a feature either wait a bit longer for more features/bugs, or make a release.

I suggest releasing without branching https://github.com/ansible/community-docs/blob/main/releasing_collections_without_release_branches.rst

Once we release a 2.0.0 (with some breaking change relative to 1.x.y), we can have a stable-1 branch so we can backport bugfixes (or even features) if needed, and release more 1.x.y versions. We currently have some deprecation removals scheduled for 2.0.0 (see #1). Maybe scheduling 2.0.0 roughly for Ansible 2.12 (i.e. next summer) would be a good idea.
(the part ^ taken from the description of https://github.com/ansible-collections/community.docker/issues/4)

@bmalynovytch @bmildren @felixfontein @Jorge-Rodriguez 

UPDATE: we [agreed](https://github.com/ansible-collections/community.mysql/discussions/537) on maintaining older (`stable*`) versions for **2 years** (backporting bugfixes, releasing, testing) since the next major version is released. Then we do final releases announcing under `major_changes` that those versions are EoL.

**Comment by Andersson007** [comment]
Questions:
1) Releasing without branches as described in the description is ok for everyone?
2) I suggest releasing version 1.0.1 on 2020-09-30 (next week Wednesday). As far as I can see, there were only fixes merged since the last release. Any thoughts?

**Comment by felixfontein** [comment]
This approach is reasonable and a good idea IMO :) For community.crypto, we eventually will have a 2.0.0 version with breaking changes (when removing something we deprecated), and then we'll add a stable-1 branch so we can still backport bugfixes (or even features, if we really want to) to release new 1.x.y versions.

I guess as long as there is no need to make a breaking change, sticking to `main` branch and versions 1.x.y is a good choice. And your users will also love that, since no breaking change means full backwards compatibility :)

**Comment by Andersson007** [comment]
community.mysql 1.0.1 has just been released https://github.com/ansible-collections/community.mysql/tags

**Comment by Andersson007** [comment]
if nobody's against, i would release 1.0.2 on 01.10.2020 (tomorrow) because of https://github.com/ansible-collections/community.mysql/pull/40

**Comment by Andersson007** [comment]
I've just released 1.0.2

**Comment by rndmh3ro** [comment]
> Release minor and patch releases whenever we want (like after adding new features or fixing bugs). Since this collection is small, there's no need to fix things in advance. 

Agreed!

> Just add features, and after a feature either wait a bit longer for more features/bugs, or make a release.

Waiting for more features/bugs seems to have no benefit, so I'd say release whenever a new feature is merged or a bug is fixed.

**Comment by Andersson007** [comment]
@rndmh3ro welcome!:)

**Comment by Andersson007** [comment]
I've just release 1.1.0

**Comment by Andersson007** [comment]
I'd like to release 1.1.1 today's evening
@Jorge-Rodriguez do you have anything to include that you almost finish but that's not yet published? We could postpone this for a couple of days.

**Comment by Andersson007** [comment]
I've just released 1.1.1

**Comment by Andersson007** [comment]
I've just released 1.1.2

**Comment by Andersson007** [comment]
`community.mysql` collection 1.2.0 has just been released.
The tarbal is available on [Galaxy](https://galaxy.ansible.com/community/mysql).
How to install it, see the [official guide](https://docs.ansible.com/ansible/latest/user_guide/collections_using.html).
The changes will also be included in the next Ansible release.

**Comment by Andersson007** [comment]
I'm going to release `community.mysql` 1.3.0 after https://github.com/ansible-collections/community.mysql/pull/62 is merged

**Comment by Andersson007** [comment]
@Jorge-Rodriguez @bmalynovytch @bmildren I've just add the following to the description:
```
Once we release a 2.0.0 (with some breaking change relative to 1.x.y), we can have a stable-1 branch so we can backport bugfixes (or even features) if needed, and release more 1.x.y versions. We currently have some deprecation removals scheduled for 2.0.0 (see #1). Maybe scheduling 2.0.0 roughly for Ansible 2.12 (i.e. next summer) would be a good idea.
(the part ^ taken from the description of ansible-collections/community.docker#4)
```
We could align the major releases with Ansble major releases as @felixfontein suggested in https://github.com/ansible-collections/community.mysql/pull/97#issuecomment-785921194.

I suggest releasing version `2.0.0` containing at least https://github.com/ansible-collections/community.mysql/pull/97 in a month before Ansible `5.0.0` - a bit more than in 6 months.

========================================================
The following major changes will be (we will update this list when needed):
**2.0.0**
```
- mysql_replication - the word ``SLAVE`` in messages returned by the module will be changed to ``REPLICA`` in ``community.mysql`` 2.0.0 (https://github.com/ansible-collections/community.mysql/issues/98).
```
**3.0.0**
```
- mysql_replication - the mode options values ``getslave``, ``startslave``, ``stopslave``, ``resetslave``, ``resetslaveall` and the master_use_gtid option ``slave_pos`` are deprecated (see the alternative values) and will be removed in ``community.mysql`` 3.0.0 (https://github.com/ansible-collections/community.mysql/pull/97).
```

**Comment by Jorge-Rodriguez** [comment]
@bmalynovytch @bmildren I'd like to move the discussion on #101 forward before we release v2

**Comment by Andersson007** [comment]
Tomorrow there will be Ansible 3.1.0 release, so I'll release this collection now to get things included

**Comment by Andersson007** [comment]
community.mysql 1.3.0 has been released

**Comment by Andersson007** [comment]
1. according to the [2.11 roadmap](https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/roadmap/ROADMAP_2_11.rst), the expected release date of 2.11 is 2021-04-26,
I suggest releasing **community.mysql 2.0.0 ~in mid-April**
2. according to the [2.12 roadmap](https://github.com/ansible/ansible/pull/74000), the expected release date of 2.12 is 2021-10-25
I suggest releasing **community.mysql 3.0.0 ~in mid-October**

The dates can be adjusted later. СС @Jorge-Rodriguez @bmalynovytch

**Comment by Jorge-Rodriguez** [comment]
LGTM, I'll have to hurry with the `REQUIRESSL` thing, so that's next on my table

**Comment by Andersson007** [comment]
FYI: As planned, I'm releasing `community.mysql 2.0.0` tomorrow morning (2021-04-15)

**Comment by Andersson007** [comment]
community.mysql 2.0.0 has been released

New branch `stable-2` has been created as well

@Jorge-Rodriguez @bmalynovytch it means that we have to backport to `stable-2` (by cherry-picking from `main`) all changes we'll merge to `main` since now (**except breaking changes!**).
I created a label `needs_backport_to_stable-2` - please put it on all PRs you'll merge since now, then create backports. Thanks!
Don't put this label on PRs with breaking changes! :)

Thanks!

**Comment by Andersson007** [comment]
So we're planning the next (`3.0.0`) major release **~in mid-October**

**Comment by Andersson007** [comment]
fyi: `stable-1` has been created as well. I've just released `community.mysql 1.4.0`. We'll backport only bug / security fixes to `stable-1`

**Comment by Andersson007** [comment]
I'm going to release community.mysql 1.4.1 and 2.1.0 later today

**Comment by Andersson007** [comment]
community.mysql 1.4.1 and 2.1.0 have been released

**Comment by Andersson007** [comment]
I'm gonna release community.mysql 2.1.1 and 1.4.2 tomorrow morning.

**Comment by Andersson007** [comment]
also there will be a dev version of 2.2.0 released tomorrow later or on Thursday

**Comment by Andersson007** [comment]
community.mysql 1.4.2 and 2.1.1 have been released

**Comment by Andersson007** [comment]
community.mysql 2.2.0-a1 has been released and available to install via galaxy or directly from https://galaxy.ansible.com/community/mysql. It contains the `mysql_role` new module. Any feedback would be much appreciated.

**Comment by Andersson007** [comment]
I'm gonna release 2.2.0 tomorrow 23.09.2021 (it's pity that nobody has tested `mysql_role` module so far, see the previous comment...)

**Comment by Andersson007** [comment]
2.2.0 (new module `mysql_role`) and 2.3.0 have just been released

**Comment by Andersson007** [comment]
community.mysql 1.4.3 has been released

**Comment by Andersson007** [comment]
I'm going to release 2.3.1 containing [this bugfix](https://github.com/ansible-collections/community.mysql/pull/233) tomorrow morning (18.10.2021).

**Comment by Andersson007** [comment]
I'm going to release 2.4.0 today and 3.0.0 later this / next week.

**Comment by Andersson007** [comment]
community.mysql 2.3.2 has been released (FYI: there won't be 2.4.0 now and in the near future)

**Comment by Andersson007** [comment]
3.0.0 containing https://github.com/ansible-collections/community.mysql/pull/252, https://github.com/ansible-collections/community.mysql/pull/244, and https://github.com/ansible-collections/community.mysql/pull/243 will be released this week

**Comment by Andersson007** [comment]
3.0.0 has been released!

@Jorge-Rodriguez @bmalynovytch @rsicart FYI I released from branches before but, in case of this collection, it turned out to be an overkill (got a bit tired to backport things to `stable-2`) as we merged the breaking changes right before the major release anyway. I would be easier to rebase breaking PRs.
I've put tag 3.0.0 in `main` and since then I will release from `main`.

It means:
- We don't need to backport things to a special release branch (it was `stable-2` but now it's irrelevant).
- We'll put next tags (3.1.0, 3.2.0, ...) in `main`.
- We'll backport bugfixes and security patches to `stable-1` (i think critical ones) and `stable-2` (all fixes).
- Of course, we don't backport new features, refactoring, etc.
- If we're gonna introduce breaking changes, we have to merge them to `main` *right before the next major release* (4.0.0). I.e. we can have PR with breaking changes open BUT **do not merge them until you release** 4.0.0! Same for all future major releases.
- If once we plan to release 4.0.0, we'll branch `stable-3` from the most recent commit before 4.0.0 release commit.

What do you think?

**Comment by rsicart** [comment]
I think I have to read docs about this collection releasing process and branches 😄 

So if I understand correctly, no breaking change PRs merged into main until next major release.

Branches like `stable-X` are long running branches for major version X?

**Comment by Andersson007** [comment]
> I think I have to read docs about this collection releasing process and branches smile
> 
> So if I understand correctly, no breaking change PRs merged into main until next major release.
> 
> Branches like `stable-X` are long running branches for major version X?

Correct:)

**Comment by Andersson007** [comment]
I'm releasing community.mysql 3.1.0, 2.3.3, and 1.4.4 this week

**Comment by Andersson007** [comment]
community.mysql 3.1.0, 2.3.3, and 1.4.4 have been released. It turned out to be a bit faster than I had expected:)

**Comment by Andersson007** [comment]
if someone else from current maintainers wants to release 3.1.1 and 2.3.4 (in other words learn how to release / become a release engineer here), I would be happy to lead you through the process. Please let me know:)

p.s. we can wait for the day / time suitable for you

**Comment by Andersson007** [comment]
I'm gonna release 3.1.1 and 2.3.4 today

**Comment by Andersson007** [comment]
[community.mysql](https://github.com/ansible-collections/community.mysql) [2.3.4](https://github.com/ansible-collections/community.mysql/blob/stable-2/changelogs/CHANGELOG.rst) and [3.1.1](https://github.com/ansible-collections/community.mysql/blob/main/changelogs/CHANGELOG.rst) have been released.

**Comment by Andersson007** [comment]
I'm gonna release 2.3.5 and 3.1.2 which will contain [critical bugfix](https://github.com/ansible-collections/community.mysql/pull/293) for the `mysql_role` module

**Comment by Andersson007** [comment]
[community.mysql](https://github.com/ansible-collections/community.mysql) versions [2.3.5](https://github.com/ansible-collections/community.mysql/blob/stable-2/changelogs/CHANGELOG.rst) and [3.1.2](https://github.com/ansible-collections/community.mysql/blob/main/changelogs/CHANGELOG.rst) have been released.

**Comment by rsicart** [comment]
I'm going to release 3.1.3 which will contain some doc improvements and 3 bugfixes:
* #307 for `mysq_user` module
* #330 for `mysql_role` module
* #336 for `mysql_replication` module

**Comment by rsicart** [comment]
community.mysql [3.1.3](https://github.com/ansible-collections/community.mysql/releases/tag/3.1.3) has been released.

**Comment by Andersson007** [comment]
i'm gonna release community.mysql 3.2.0, 2.3.6 and 1.4.5 today

**Comment by Andersson007** [comment]
The [community.mysql](https://github.com/ansible-collections/community.mysql) collection [1.4.5](https://github.com/ansible-collections/community.mysql/blob/stable-1/CHANGELOG.rst), [2.3.6](https://github.com/ansible-collections/community.mysql/blob/stable-2/CHANGELOG.rst) and [3.2.0](https://github.com/ansible-collections/community.mysql/blob/main/CHANGELOG.rst) versions have been released!

**Comment by Andersson007** [comment]
I'm gonna release versions 1.4.7, 2.3.8 and 3.3.0 today

**Comment by Andersson007** [comment]
The [community.mysql](https://galaxy.ansible.com/community/mysql) collection versions [1.4.7](https://github.com/ansible-collections/community.mysql/blob/stable-1/CHANGELOG.rst), [2.3.8](https://github.com/ansible-collections/community.mysql/blob/stable-2/CHANGELOG.rst) and [3.3.0](https://github.com/ansible-collections/community.mysql/blob/main/CHANGELOG.rst) have been released!

Great job @betanummeric @rsicart @hubiongithub @moledzki @bmalynovytch, thank you!

**Comment by laurent-indermuehle** [comment]
One month has already passed since the last release. No fragments exist and the last PRs were about the CI. So I will skip this month's release.

**Comment by Andersson007** [comment]
> One month has already passed since the last release. No fragments exist and the last PRs were about the IC. So I will skip this month's release.

@laurent-indermuehle yep, thanks for taking care of this!

**Comment by Andersson007** [comment]
I'm going to release 3.5.0 today.

**Comment by Andersson007** [comment]
Copied from Bullhorn update: The [community.mysql](https://github.com/ansible-collections/community.mysql) collection version [3.5.0](https://github.com/ansible-collections/community.mysql/blob/main/CHANGELOG.rst#v3-5-0) has been released! Special thanks to [rsicart](https://github.com/rsicart), [betanummeric](https://github.com/betanummeric), [laurent-indermuehle](https://github.com/laurent-indermuehle) and [aneustroev](https://github.com/aneustroev)!

**Comment by rsicart** [comment]
I'm going to release 3.5.1 today.

**Comment by rsicart** [comment]
The community.mysql collection version [3.5.1](https://github.com/ansible-collections/community.mysql/blob/main/CHANGELOG.rst#v3-5-1) has been released! Special thanks to [andersson007](https://github.com/Andersson007), [laurent-indermuehle](https://github.com/laurent-indermuehle) and [betanummeric](https://github.com/betanummeric)!

**Comment by laurent-indermuehle** [comment]
I'm going to release 1.5.1, 2.4.1 and 3.7.0 today.

**Comment by Andersson007** [comment]
I'm going to release 3.7.1 today.

**Comment by Andersson007** [comment]
the [community.mysql](https://galaxy.ansible.com/community/mysql) collection version [3.7.1](https://github.com/ansible-collections/community.mysql/releases/tag/3.7.1) has been released!

**Comment by Andersson007** [comment]
I'm going to release 3.7.2 today.
UPDATE: Done

**Comment by laurent-indermuehle** [comment]
I'm going to release 3.8.0 today.
UPDATE: Done

**Comment by Andersson007** [comment]
Moving to forum, closing this one. Please join the https://github.com/ansible-collections/community.mysql/pull/589 and let's continue on the forum, thanks everyone!
Next time when releasing, please open a permanent forum topic called something like ` The community.mysql collection release plan & announcements` tagged with `mysql` under `Project Discussions` as a replacement for this issue.
